### DISCUSSION TITLE
Real-time STT delay became longer.

### DISCUSSION THREAD

**Discussion by danieldev0724** [deepgram/community#336] [General help] [answered]
Hi.
I faced a problem of real-time STT delay while using Deepgram.
I used Python FastAPI for backend and use websockets to get audio stream from frontend and also to get transcription using Deepgram API.

This is my code for creating Deepgram socket.

```
from fastapi import FastAPI, Request, WebSocket
from typing import Dict, Callable
from deepgram import Deepgram
import json
import asyncio
from config import DEEPGRAM_API_KEY

dg_client = Deepgram(DEEPGRAM_API_KEY)

async def process_audio(sender_socket: WebSocket, receiver_socket: WebSocket, sender: str):
    async def get_transcript(data: Dict) -> None:
        if 'channel' in data:
            transcript = data['channel']['alternatives'][0]['transcript']
            if transcript:
                print("n\n\ntranscript: ", transcript)
                msg_data = json.dumps({"sender": sender, "transcript": transcript, "request": "customer_new_audio_sent"})
                await sender_socket.send_text(transcript)
                await receiver_socket.send_text(transcript)
    deepgram_socket = await connect_to_deepgram(get_transcript)
    return deepgram_socket

async def connect_to_deepgram(transcript_received_handler: Callable[[Dict], None]):
    try:
        print("deepgram socket created")
        socket = await dg_client.transcription.live({'punctuate': True, 'interim_results': False, 'utterance_split': False})
        socket.registerHandler(socket.event.CLOSE, lambda c: print(f'Connection closed with code {c}.'))
        socket.registerHandler(socket.event.TRANSCRIPT_RECEIVED, transcript_received_handler)        
        async def send_keep_alive():
            while True:
                if socket is not None:
                    await asyncio.sleep(3)
                    socket.send('{"type": "KeepAlive"}')
        asyncio.create_task(send_keep_alive())
        return socket
    except Exception as e:
        raise Exception(f'Could not open socket: {e}')
```

It works fine at first. However, if I stop the recording and start streaming the audio recording again, Deepgram starts responding after 1-3 minutes, even 7 minutes.
I don't know why. Please help me.

**Comment by jjmaldonis** [comment]
Hey @danieldev0724, I've never heard of this behavior before. Are you able to reproduce the issue without using FastAPI? If so, can you please post the full code so we can reproduce it?

**Reply by danieldev0724** [reply]
Hi
Thanks for your reply.
But I have to use FastAPI and the project is a little big.
As I mentioned, it works fine at first. If I stop and restart audio record streaming at frontend, the bytes are sent to backend and fed into deepgram socket. But the deepgram starts responding after 1 minute.

**Reply by danieldev0724** [reply]
I tried for react, but the result is the same.
This is the react code.

```
useEffect(() => {
  const deepgramLive = new WebSocket(`wss://api.deepgram.com/v1/listen`, [
    'token',
    deepgramApiKey,
  ]);
  
  deepgramLive.onclose = () => {
    console.log("Connection closed.");
  };
  
  deepgramLive.onmessage = async (message) => {
    const received = JSON.parse(message.data);
    console.log(received)
    if (received.channel) {
      const transcript = received.channel.alternatives[0].transcript;
      if (transcript) {
        console.log(transcript);
      }
    }          
  };
  
  setInterval(() => {
    if (deepgramLive && deepgramLive.readyState === WebSocket.OPEN) {
      deepgramLive.send(JSON.stringify({ type: 'KeepAlive' }));
    }
  }, 5000);
  
  setAudwebsckt(deepgramLive);
}, []);

const onStartRecording = async () => {
    setRecordingStatus("recording");
    const media = new MediaRecorder(stream, {
      mimeType: "audio/webm",
    });

    mediaRecorder.current = media;
    mediaRecorder.current.start(250);
    mediaRecorder.current.addEventListener("dataavailable", async (event) => {
      if (event.data.size > 0 && audwebsckt.readyState == 1) {
        audwebsckt.send(event.data);
        console.log("sent audio data");
      }
    });
  };
```

It works fine at first. But if I stop and restart audio record streaming, the deepgram responses after 1 minute.

**Reply by jjmaldonis** [reply]
Is the start/stop functionality you implemented creating a new websocket? The easiest way to check is to watch for `metadata` messages and inspect the request ID in the metadata JSON. If multiple websockets are being created, that is likely not what you want to do.

The `setInterval` is likely to be the culprit. The interval may continue to run even when you don't want it to. You may need to carefully manage the state of this interval.

**Reply by danieldev0724** [reply]
In react code, there will be only one socket. I set the socket using useState.
setInterval is responsible for sending keep-alive status. So I think I have to use setInterval.

**Reply by jjmaldonis** [reply]
Can you test what happens if you completely remove the `setInterval` (and therefore don't send any KeepAlive messages)? If (for this test) you talk constantly, the KeepAlive functionality won't be necessary. Then, without the KeepAlive code, stop and restart the audio recording. Does the 1 minute delay in response still happen?

**Reply by danieldev0724** [reply]
At first of dev, I didn't use KeepAlive. If I don't use KeepAlive, the socket will be dead when stop talking for a while.
As you know, while a person is talking, there will be some intervals.
Because of this problem, I used KeepAlive func.

**Comment by jpvajda** [comment, marked as answer]
This discussion has moved to Discord https://discord.com/channels/1108042150941294664/1148508501358948433
